### ISSUE TITLE
Support for explicit REJECT action in policy rules

### ISSUE THREAD

**Issue by zetsub0u** [projectcalico/calico#4500]
Hi,

Quick question, I don't see a way to set the policy for the rules from DROP to REJECT (silently drop vs quick RST), is this accurate? If so, any reason why this isn't configurable and would it be acceptable PR?

Thanks.

**Comment by lwr20** [comment]
Its always been DROP, mostly because in the firewall world, its a bad idea to let people know that the IP address has a host behind it, and to slow down their reconnaissance of your hosts.

What's your use case for REJECT?

**Comment by zetsub0u** [comment]
I know and i agree completely, but we've had our firewalls do REJECT for years, on our custom implementation in docker and after that with kube-router (they do reject too) and when we switched to DROP a lot of users started complaining because they used the behavior of the quick connection refused to debug network policy issues for example.
We are trying to convince them it's better to go DROP but we might lose the battle, so i was wondering if having the option somewhere would be an acceptable change.

**Comment by lwr20** [comment]
This almost got added previously:  https://github.com/projectcalico/felix/pull/2161/files

**Comment by olljanat** [comment]
Any update to this one? Our use case is that we want to go zero trust kind of architecture where we block all connections between namespaces are blocked by default and REJECT is just more clear for those who are trying to debug why some of needed connections fails.

Example I'm new with Kubernetes and just wasted couple of hours because I though that there was some routing, etc issue on my Kubernetes/infra setup before remembered that there is also that default network policy which blocks everything...

**Comment by olljanat** [comment]
Btw. I looks to be that https://github.com/projectcalico/felix/pull/2532 already added support for REJECT at least on some parts of code https://github.com/projectcalico/felix/blob/87c68a8b66f6a57170a3ab5084c8c7b66baacdd4/iptables/actions.go#L89-L99 but not sure (at least yet) that what is actually still missing.

**Comment by olljanat** [comment]
@lwr20 any possibility to get this to core team list or at least get some guidance that what would be right way and place to implement this one?

**Comment by lwr20** [comment]
@caseydavenport is the core team lead - perhaps he can comment?

**Comment by caseydavenport** [comment]
Yep. I think we're still looking for someone from the community to pick this thread up. I agree the last PR was close - probably needed some work on the tests mostly.

**Comment by olljanat** [comment]
Ok. Created PR #5735 based on code from https://github.com/projectcalico/felix/pull/2161 and added initial set of tests for now and will try to working with it when I have spare time.

**Comment by olljanat** [comment]
Status update. #5735 have been there for a while and should contain most of the needed logic.

However it looks that I will need help from someone who know this codebase better to understand that why `make test` in felix folder fails? For me it looks that felix does not actually read `DropActionOverride` setting value even when it is configured.

**Comment by olljanat** [comment]
Status update.  #5735 have been ready for code review about one month now. I only have rebased it time to time when other merged PRs has required it.

**Comment by james-callahan** [comment]
While #5735 is merged, it's not quite what I wanted/expected with this issue.
I was hoping that [`action` in a Rule](https://docs.tigera.io/calico/latest/reference/resources/globalnetworkpolicy#rule) would gain a `Reject` value which could be used instead of `Deny`.

**Comment by caseydavenport** [comment]
Updated the title to represent the remaining request scope

**Comment by james-callahan** [comment]
Can this be reopened? I still think it would be useful

**Comment by ViliusS** [comment]
40 year later, and the myth of "DROP is better than REJECT because it _hides_ IP address" still persists, and apparently implemented as a default mode by major Kubernetes networking plugins :)

Anyway, does anybody know if changes implemented in https://github.com/projectcalico/calico/pull/5735 can be applied when using Tigera Operator? I only can find how to change Felix configuration using Helm, but not with Tigera.

**Comment by caseydavenport** [comment]
@ViliusS yep, you can just specify the value in the `projectcalico.org/v3` `FelixConfiguration` API - e.g., "kubectl edit" or using Argo / etc., to create it like any other resource.

**Comment by ViliusS** [comment]
But I assume it will be overwritten by operator at will? I see that currently FelixConfiguration object is managed by operator.

**Comment by caseydavenport** [comment]
It shouldn't - operator only sets a select handful of fields, and should refuse to overwrite user config (if you see it doing otherwise, that is likely a bug!)

**Comment by ViliusS** [comment]
Ah, nevermind. I see that Calico on GKE fails when configured with `iptablesFilterDenyAction: REJECT`. I can only find a mixed data on that. Some say Google's Container OS does not include iptables. Some say iptables are there but are limited. I guess I'm suck with DROP for foreseeable future.

Anyway, thank you for your help @caseydavenport

**Comment by caseydavenport** [comment]
Interesting, I wasn't aware of limited iptables support on GKE - that's surprising! 

Wonder if we can / need to do anything here for an nftables equivalent.

**Comment by ViliusS** [comment]
I tried to find if they use nftables or something else, but could not find concrete evidence. Google's documentation on Calico, if if can be customized, and if it would be considered supported configuration is also bleak.

As I don't have an active Google support contract I don't know how to investigate this further. All I can do is share error logs from GKE calico-nodes when switching to REJECT.

P.S. Is there a better way to discuss this? I feel like this is getting out of scope of this issue.

**Comment by caseydavenport** [comment]
Slack would probably be best! Or feel free to open another issue / discussion (https://github.com/orgs/projectcalico/discussions) and tag me on it!
